homeassistant/components/modbus/modbus.py

```diff
@@ -169,43 +169,52 @@ async def async_stop_modbus(event: Event) -> None:
 
     hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_stop_modbus)
 
+    UNIT = "unit"
+
     async def async_write_register(service: ServiceCall) -> None:
         """Write Modbus registers."""
-        slave = 1
-        if ATTR_UNIT in service.data:
-            slave = int(float(service.data[ATTR_UNIT]))
-
-        if ATTR_SLAVE in service.data:
-            slave = int(float(service.data[ATTR_SLAVE]))
-        address = int(float(service.data[ATTR_ADDRESS]))
+        # Make a log message, in case some very old automations still use unit=
+        if UNIT in service.data:
+            _LOGGER.error(
+                "Be advised: unit= is no longer supported, please use device_address="
+            )
+            service.data[ATTR_DEVICE_ADDRESS] = service.data[UNIT]
+        device_id = service.data.get(
+            ATTR_DEVICE_ADDRESS, service.data.get(ATTR_SLAVE, 1)
+        )
+        address = service.data[ATTR_ADDRESS]
         value = service.data[ATTR_VALUE]
         hub = hub_collect[service.data.get(ATTR_HUB, DEFAULT_HUB)]
         if isinstance(value, list):
             await hub.async_pb_call(
-                slave,
+                device_id,
                 address,
                 [int(float(i)) for i in value],
                 CALL_TYPE_WRITE_REGISTERS,
             )
         else:
             await hub.async_pb_call(
-                slave, address, int(float(value)), CALL_TYPE_WRITE_REGISTER
+                device_id, address, int(float(value)), CALL_TYPE_WRITE_REGISTER
             )
 
     async def async_write_coil(service: ServiceCall) -> None:
         """Write Modbus coil."""
-        slave = 1
-        if ATTR_UNIT in service.data:
-            slave = int(float(service.data[ATTR_UNIT]))
-        if ATTR_SLAVE in service.data:
-            slave = int(float(service.data[ATTR_SLAVE]))
+        # Make a log message, in case some very old automations still use unit=
+        if UNIT in service.data:
+            _LOGGER.error(
+                "Be advised: unit= is no longer supported, please use device_address="
+            )
+            service.data[ATTR_DEVICE_ADDRESS] = service.data[UNIT]
+        device_id = service.data.get(
+            ATTR_DEVICE_ADDRESS, service.data.get(ATTR_SLAVE, 1)
+        )
```

Comment:
Please reconsider my suggestion:
- you should not log an error, as it still works (it should be a warning)
- you should tell users when it will break (eg. will be removed fully in 2026.4)

I don't think you should have two "good" ways to do it. If `slave` is a fallback it should be deprecated.

Reply:
Just because it was accepted some time in the past, it doesn't mean it was the best way of doing things.
We were a lot more leniant about breaking changes in the past, when release notes were considered sufficient.

Now the policy is to warn users actively when possible, and give them a timeframe.

If you think the term "slave" is acceptable, then no point removing it.
If it's not acceptable, then we should actively tell people to change it, and log a warning accordingly.